### PR TITLE
Rebase or merge, pick your poison

### LINKED ISSUE #43: New architecture
The current approach to handling webhooks is fundamentally flawed. We're handling every kind of event differently (e.g. we could find the pipeline by MR or the MR by pipeline), but the only things we care about is whether *something* of significance happened - a MR was opened, a commit was made, a pipeline changed status. The handling of significant events should be identical at all times.

A very important reason why that has to be the case is that events in one MR may trigger actions in other MRs - for example, merging MR `#1` may result in rebasing MR `#2`, which will trigger a pipeline, whose change of status to successful should trigger a merge of MR `#2`.

~I think webhook events should only provide information about which project was impacted. Currently the hooks that come to mind are Pipeline, MR and Push hooks (Build is redundant as we need to wait for a pipeline anyways).~ All webhook events are supported now, as long as `project` is specified. We fetch all open MRs and try to find actions for each of them.

For each such project, we will:

1. Ensure it's supported by pitgull (for now this isn't strictly necessary, but later we'll see if it has a configuration file and optionally if it's registered if we're launching to the public)
2. Get all of its merge requests
3. If a MR matches a rule, put that MR on a per-project global queue. In another fiber, changes to that queue (or MR-related triggers) will be watched and an attempt to merge *exactly one* MR will be made. If the project is configured to require rebasing other MRs, we can then rebase the next suitable MR in sequence. If we don't need to rebase before merging others - we should probably wait for the master pipeline somehow, or re-trigger the next MR's pipeline. TODO.

### PR DESCRIPTION
**Pull request by kubukoz**
Closes #43.

~For now, this is targetting #83, but will be rebased to main once #83 is done.~

### CODE REVIEW

**Comment by kubukoz** [comment]
Preliminary results from the test repository, merged while Scala Steward was working :) 

image

Every one was rebased exactly once before being merged (the repo rules enforced semilinear history).

only one out of 7 wasn't merged because of conflicts:

image

After I fixed this, pitgull was able to proceed with merging that one as well.

**Comment by kubukoz** [comment]
Blocked by https://github.com/valskalla/odin/issues/229, kind of. Or I can get back to the logging part later and move it somewhere else for now.

**Comment by majk-p** [comment]
> Blocked by [valskalla/odin#229](https://github.com/valskalla/odin/issues/229), kind of. Or I can get back to the logging part later and move it somewhere else for now.

How about manually creating the file for now, or maybe just mentioning it in some _install instructions_ section? Let's go with either of those and LGTM

**Comment by kubukoz** [comment]
The file logger is disabled right now, i guess we can leave it like that for the time being

File: src/main/scala/io/pg/actions.scala [hunk 1]
```diff
@@ -59,67 +69,78 @@ object ProjectActions {
...
-    case Matcher.Description(text) => descriptionMatches(text)
+  val compileMatcher: Matcher => MatcherFunction[MergeRequestState] = {
+    case Matcher.Author(email)          => autorMatches(email)
+    case Matcher.Description(text)      => descriptionMatches(text)
     case Matcher.PipelineStatus(status) => statusMatches(status)
-    case Matcher.Many(values) => values.foldMapK(compileMatcher)
+    case Matcher.Many(values)           => values.foldMapK(compileMatcher)
   }
 
+  // def compile(
```
Comment (kubukoz) [inline src/main/scala/io/pg/actions.scala:126]: Todo

**Pull request merged**